Python/Module6_Testing/Hypothesis.md

```diff
@@ -0,0 +1,1047 @@
+
+#### `st.booleans ()`
+
+[st.booleans()](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.booleans) generates either `True` or `False`. This strategy will shrink towards `False`
+
+```python
+>>> st.booleans().example()
+False
+```
+<!-- #endregion -->
+
+<!-- #region -->
+
+#### `st.lists ()`
+
+[st.lists](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.lists) accepts *another* strategy, which describes the elements of the lists being generated. You can also specify:
+ - bounds on the length of the list
+ - if we want the elements to be unique
+ - a mechanism for defining "uniqueness"
+
+For example, the following strategy describes lists whose length varies from 2 to 10, and whose entries are integers on the domain $[-10, 20]$:
+
+```python
+>>> st.lists(st.integers(-10, 20), min_size=2, max_size=10).example()
+[-10, 0, 5]
+```
+
+**`st.lists(...)` is the strategy of choice anytime we want to generate sequences of varying lengths with elements that are, themselves, described by strategies**.
+<!-- #endregion -->
+
+<!-- #region -->
+#### `st.floats()`
+
+[st.floats](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.floats) is a powerful strategy that generates all variety of floats, including `math.inf` and `math.nan`. You can also specify:
+ - whether `math.inf` and `math.nan`, respectively, should be included in the data description
+ - bounds (either inclusive or exclusive) on the floats being generated; this will naturally preclude `math.nan` from being generated
+ - the "width" of the floats; e.g. if you want to generate 16-bit or 32-bit floats vs 64-bit
+   (while Python's `float` is (usually) 64-bit, `width=32` ensures that the generated values can
+   always be losslessly represented in 32 bits.  This is mostly useful for NumPy arrays.)
+
+For example, the following strategy 64-bit floats that reside in the domain $[-100, 1]$:
+
+```python
+>>> st.floats(-100, 1).example()
+0.3670816313319896
+```
+<!-- #endregion -->
+
+<!-- #region -->
+#### `st.tuples()`
+
+The [st.tuples](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.tuples) strategy accepts $N$ Hypothesis strategies, and will generate length-$N$ tuples whose elements are drawn from the respective strategies that were specified as inputs.
+
+For example, the following strategy will generate length-3 tuples whose entries are: integers, booleans, and floats:
+
+```python
+>>> st.tuples(st.integers(), st.booleans(), st.floats()).example()
+(4628907038081558014, False, -inf)
+```
+<!-- #endregion -->
+
+<!-- #region -->
+#### `st.text()`
+
+The [st.text](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.text) accepts an "alphabet" – a collection of string-characters – from which it will construct strings of varying lengths, whose bounds can be specified by the user.
```

Comment:
The alphabet argument must be a collection of length-one strings (note that a string is such a collection), or a strategy for generating such values (eg st.characters()).